### ISSUE TITLE
Einzelne Spendenbescheinigungen ausstellen funktioniert nicht

### ISSUE THREAD

**Issue by JPT77** [openjverein/jverein#154]
Mitglied auswählen, Kontextmenü, Spendenbescheinigung.
Resultat: Kein Mitgliedskonto ausgewählt
`
Fri Feb 16 16:06:09 CET 2024][DEBUG][pool-1-thread-1][de.willuhn.jameica.messaging.NamedConcurrentQueue.deliver] deliver message [Fri Feb 16 16:06:09 CET 2024][ERROR] Kein Mitgliedskonto ausgewählt`

Spendenbescheinigung/neu (manuell) 
Resultat: Kein Mitgliedskonto ausgewählt. 
Was irgendwie keinen Sinn macht, weil man in dieser Übersicht kein Mitgliedskonto auswählen kann.

`[Fri Feb 16 16:39:28 CET 2024][DEBUG][pool-1-thread-1][de.willuhn.jameica.messaging.NamedConcurrentQueue.deliver] deliver message [Fri Feb 16 16:39:28 CET 2024][ERROR] Kein Mitgliedskonto ausgewählt
`

Bezieht sich  Mitgliedskonto auf die Tabelle "Mitgliedskonto" oder auf "Mitglied"?

Kann es was mit den Abrechnungsläufen zu tun haben?

Bitte mehr logging. Es scheint ja schon auf DEBUG zu stehen, aber trotzdem kommt nix...

**Comment by JohannMaierhofer** [comment]
Der Fehler ist bereits mit Issue #75 bekannt und wurde behoben. Bitte auf die neue Version 2.8.21 umsteigen.

**Comment by JPT77** [comment]
Ah, super. Ja, den Dialog für die manuelle Anlage einer Spendenbescheinigung kann ich nun öffnen.
Leider will JVerein aber für Mitglieder, bei denen im Laufe des Jahres 23 nur jeweils eine Mitgliedsbeitragsbuchung vorliegt, nach wie vor nicht automatisch Spendenbescheinigungen anlegen. 
Und bei den manuell anzulegenden Spendenbescheinigungen erscheinen keine Einträge in der Liste der Zahlungen.

**Comment by JohannMaierhofer** [comment]
> Ah, super. Ja, den Dialog für die manuelle Anlage einer Spendenbescheinigung kann ich nun öffnen. Leider will JVerein aber für Mitglieder, bei denen im Laufe des Jahres 23 nur jeweils eine Mitgliedsbeitragsbuchung vorliegt, nach wie vor nicht automatisch Spendenbescheinigungen anlegen. Und bei den manuell anzulegenden Spendenbescheinigungen erscheinen keine Einträge in der Liste der Zahlungen.

Ich denke, den letzten Punkt habe ich gerade mit dem Pull Request #158 gelöst.

**Comment by JohannMaierhofer** [comment]
Hallo @JPT77 , ich kann das Problem warum das automatische Anlegen nicht gehen sollte nicht nachvollziehen. Bei mir geht es. Es darf erstens noch keine Spendenbescheinigung für die Buchungen generiert worden sein. Dann muss bei jedem Mitglied eine Strasse, PLZ und Ort eingetragen sein. Natürlich muss auch in der Buchungsart das Flag Spende gesetzt sein.
Bitte mal prüfen ob das der Fall ist.

**Comment by JPT77** [comment]
Ah, möglicherweise liegt es am Flag Spende. 
Aber soweit ich das verstanden habe, sind Mitgliedsbeiträge steuerlich äquivalent zu Spenden, wenn der Verein denn als gemeinnützig eingestuft ist durch das Finanzamt?

**Comment by JohannMaierhofer** [comment]
Ja, bei gemeinnützigen Vereinen können für Mitgliedsbeiträge Spendenbescheinigungen ausgestellt werden. Bei meinem ist das so. Du musst also dann unter Administration->Einstellungen->Buchungsarten bei der Buchungsart für Beiträge den Haken Spende setzten.

**Comment by JPT77** [comment]
Das ist gesetzt. Es werden auch Mitgliedsbeiträge mit einbezogen.
Aber gerade wenn nur ein oder zwei Buchungen pa existieren, dann funktioniert es seltsamerweise nicht. 
Ist jetzt nicht so kritisch, bei Minibeträgen braucht man eh keine Spendenbescheinigung. und in Einzelfällen kann ich es ja auch manuell machen. 

Danke, dass du dich hier engagierst. Ich wollte eigentlich auf WISO umsteigen, da mir JVerein recht inaktiv vorkam. 
Unter diesen Umständen würde ich doch gerne bleiben, vor allem aus Überzeugung weg OpenSource. Ich bin auch Java-Developer und kann meinen Teil beitragen. Nur alleine in weiter Flur hätte ich das nicht gemacht.

**Comment by JohannMaierhofer** [comment]
Weil du gerade von Minibeiträgen redest fällt mir noch was ein.
Erstens musst du unter Administration->Einstellungen->Spendenbescheinigungen den Eintrag "Mitgliedsbeiträge dürfen bescheinigt werden" anklicken bei dem gemeinnützigen Verein. Das ist nur für den Ausdruck nötig.
Dann gibt es dort auch noch das Feld Mindestbeitrag. Der Betrag in der Spendenbescheinigung muss größer oder gleich sein als der Wert. Ansonsten wird keine Bescheinigung generiert. Am besten setzt du es auf 0, dann sollte für alle Beträge etwas generiert werden.

**Comment by JPT77** [comment]
Ah, ja es erscheint jetzt mehr, aber nicht alles. etwa 30 buchungen erscheinen, 55 sollten es aber sein.
`select * from buchung where spendenbescheinigung is null and not mitgliedskonto is null and datum < "2023-01-01";`
Verzicht ist für alle false, Buchungsart 55 oder 56 bzw Konto 2110 und 2120 (Mitgliedsbeitrag < oder > 300). Sonst sehe ich nichts relevantes. 

Könnte man solche impliziten Filter als Readonly Felder in der entsprechenden Ansicht hinzufügen?
Also dass da beispielsweise steht "es werden nur Beträge die 50€ pa übersteigen angezeigt", mit Referenz wo man das verstellen kann?
ich denke, da gibt es viele solcher Situationen... die alle zu finden ist natürlich schwierig.

**Comment by JohannMaierhofer** [comment]
Ich habe deine Anregung aufgenommen und werde einen Infotext einbauen.
Ich verstehe aber immer noch nicht wo das Problem ist. Ist in der Buchungsliste den Buchungen auch bei jeder Buchung der Art Spende ein Mitglied zugeordnet worden und ist das Datum der Buchung im selektierten Jahr?

**Comment by JPT77** [comment]
Ja, deswegen habe ich ja das SQL statement oben. daraus gehen die drei requirements hervor: es wurde noch keine Bescheinigung erstellt, es ist ein Mitgliedskonto zugeordnet und das Datum liegt in 2022.

ich habe eben nochmal geprüft. zu jeder referenzierten MK ID existiert auch ein MK. Das B.DATUM wie auch das MK.DATUM liegt jeweils im Jahr 2022
oder hast du einen anderen vorschlag für das SQL?

**Comment by JohannMaierhofer** [comment]
Also, der Code macht folgendes:
Um ein Mitglied zu suchen
"SELECT mitglied.id, sum(buchung.betrag) " + "FROM buchung "
        + "  JOIN buchungsart ON buchung.buchungsart = buchungsart.id "
        + "  JOIN mitgliedskonto ON buchung.mitgliedskonto = mitgliedskonto.id "
        + "  JOIN mitglied ON mitgliedskonto.mitglied = mitglied.id "
        + "WHERE year(buchung.datum) = ? " + "  AND buchungsart.spende = true "
        + "  AND buchung.spendenbescheinigung IS NULL "
        + "  AND buchung.mitgliedskonto IS NOT NULL "
        // rdc: Nur Mitglieder mit bekannter Adresse
        + "  AND (mitglied.strasse IS NOT NULL OR LENGTH(mitglied.strasse) > 0) "
        + "  AND (mitglied.plz IS NOT NULL OR LENGTH(mitglied.plz) > 0) "
        + "  AND (mitglied.ort IS NOT NULL OR LENGTH(mitglied.ort) > 0) "
        + "GROUP BY mitglied.name, mitglied.vorname, mitglied.id "
        // rdc: Nur Spendenbescheinigungen, deren Betrag >= Mindestbetrag
        + "HAVING sum(buchung.betrag) >= ? "
        + "ORDER BY mitglied.name, mitglied.vorname, mitglied.id";

Um dann die Buchungen für das Mitglied zu suchen:
"SELECT buchung.id, buchung.datum FROM buchung "
        + "  JOIN buchungsart ON buchung.buchungsart = buchungsart.id "
        + "  JOIN mitgliedskonto ON buchung.mitgliedskonto = mitgliedskonto.id "
        + "WHERE year(buchung.datum) = ? " + "  AND buchungsart.spende = true "
        + "  AND mitgliedskonto.mitglied = ? "
        + "  AND buchung.spendenbescheinigung IS NULL "
        + "  AND buchung.mitgliedskonto IS NOT NULL "
        + "ORDER BY buchung.datum";

**Comment by JohannMaierhofer** [comment]
Aber in deiner Abfrage fehlt doch ob die Art der Buchungsart der Buchung eine Spende ist. Das habe ich oben schon erwähnt.
![Screenshot_3](https://github.com/openjverein/jverein/assets/126261667/729a6a69-6d4e-4bc3-a621-f4a9887be808)

**Comment by JPT77** [comment]
Na, ich habe doch das Statement aus JVerein ausgeführt, dass du  mir geschickt hast. ich habe nur 2022 und 0 eingesetzt.

**Comment by JohannMaierhofer** [comment]
Möglicherweise habe ich etwas gefunden. Im Code nach dem Mitglieds Query wird nochmals abgefragt ob Straße, PLZ und Ort nicht null sind und nicht Länge 0. Anscheinend hat der Entwickler schon nicht passende Einträge bekommen.
Nun glaube ich, dass das Query einen Fehler hat.
In:
+ " AND (mitglied.strasse IS NOT NULL OR LENGTH(mitglied.strasse) > 0) "
+ " AND (mitglied.plz IS NOT NULL OR LENGTH(mitglied.plz) > 0) "
+ " AND (mitglied.ort IS NOT NULL OR LENGTH(mitglied.ort) > 0) "
müsste statt OR eigentlich ein AND stehen. Die Attribute dürfen nicht NULL sein und die Länge muß > 0 sein.
Wenn man also nur das Query ausführt bekommt man auch Datensätze wo diese Attribute die Länge 0 haben. Der Code filter diese aber dann wieder aus. Darum die Diskrepanz.
Die Frage wäre jetzt ob bei deiner der drei genannten Mitglieder eines dieser drei Attribute eine Länge 0 hat, also die Felder leer sind.

**Comment by JPT77** [comment]
Ah, verdammt. ich dachte, ich hätte  nachgeguckt. aber ich war beim falschen.
Die drei haben tatsächlich keine Adresse.

Ja, du hast recht. es sollte NOT NULL **AND** LEN sein. 
Gut, dann wär das geklärt. Man muss nur hartnäckig sein. 
Auf eine erfolgreiche Zusammenarbeit! :)

**Comment by JohannMaierhofer** [comment]
> Für die Erkennbarkeit der Ursache wär aber besser, wenn JVerein Platzhalter druckt statt garnichts zu machen.
> 
> 
Ja, siehe #161
